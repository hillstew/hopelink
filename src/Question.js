import React, { Component } from "react";
import AnswerCard from "./AnswerCard";

class Question extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentAnswer: "",
      isSubmitted: false,
      currentQuestionIndex: 0
    };
  }

  updateCurrentAnswer = event => {
    this.setState({ currentAnswer: event.target.value });
  };

  toggleSubmitAnswer = event => {
    event.preventDefault();
    this.setState({ isSubmitted: true });
  };

  changeToNextQuestion = () => {
    let nextQuestionIndex = this.state.currentQuestionIndex + 1;
    if (nextQuestionIndex === 20) {
      nextQuestionIndex = 0;
    }
    this.setState({
      currentQuestionIndex: nextQuestionIndex
    });
  };

  showQuestion = () => {
    this.setState({ isSubmitted: false });
  };

  changeBackToPreviousQuestion = () => {
    let previousQuestionIndex = this.state.currentQuestionIndex - 1;
    if (previousQuestionIndex === 0) {
      previousQuestionIndex = 19;
    }
    this.setState({
      currentQuestionIndex: previousQuestionIndex
    });
  };

  render() {
    const {
      showSavedAnswers,
      savedAnswers,
      updateSavedAnswers,
      techQuestions
    } = this.props;
    const {
      isSubmitted,
      currentAnswer,
      currentQuestionIndex
    } = this.state;
    if (showSavedAnswers && savedAnswers.length >= 1) {
      return savedAnswers.map(answer => {
        return (
          <AnswerCard
            currentQuestion={answer.question.question}
            currentAnswer={answer.answer}
            updateSavedAnswers={updateSavedAnswers}
            techQuestions={techQuestions}
            currentQuestionIndex={currentQuestionIndex}
            savedAnswers={savedAnswers}
            showSavedAnswers={showSavedAnswers}
          />
        );
      });
    } else if (isSubmitted) {
      return (
        <AnswerCard
          currentAnswer={currentAnswer}
          updateSavedAnswers={updateSavedAnswers}
          techQuestions={techQuestions}
          currentQuestionIndex={currentQuestionIndex}
          isSubmitted={isSubmitted}
          showQuestion={this.showQuestion}
          savedAnswers={savedAnswers}
          showSavedAnswers={showSavedAnswers}
        />
      );
    } else {
      return (
        <div className="question-div">
          <h2>{techQuestions[currentQuestionIndex].question}</h2>
          <form onSubmit={this.toggleSubmitAnswer}>
            <textarea
              type="text"
              value={this.state.value}
              onChange={this.updateCurrentAnswer}
            />
            <input
              type="submit"
              value="Submit"
              className="question-submit-input"
            />
          </form>
          <button
            onClick={this.changeBackToPreviousQuestion}
            className="next-back-btn"
          >
            BACK
          </button>
          <button onClick={this.changeToNextQuestion} className="next-back-btn">
            NEXT
          </button>
        </div>
      );
    }
  }
}

export default Question;
