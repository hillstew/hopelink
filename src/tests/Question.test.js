import React from "react";
import { shallow } from "enzyme";
import Question from "../Question";

let techQuestions = [
  {
    question:
      "What is the difference between local and global variables, and how do they relate to hoisting?",
    category: "javascript",
    link:
      "https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/",
    video: "https://www.youtube.com/watch?v=iJKkZA215tQ&t=244s"
  },
  {
    question: "What is closure?",
    category: "javascript",
    link:
      "https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-closure-b2f0d2152b36",
    video: "https://www.youtube.com/watch?v=-xqJo5VRP4A"
  },
  {
    question: "Describe private and public class fields",
    category: "javascript",
    link:
      "https://tylermcginnis.com/javascript-private-and-public-class-fields/",
    video: "https://www.youtube.com/watch?v=ZxpdR_C7RVA"
  },
  {
    question: "What is the difference between forEach and map?",
    category: "javascipt",
    link:
      "https://medium.com/@bretdoucette/foreach-v-map-javascript-the-definition-and-the-difference-5eaa65e15b65",
    video: "https://codeburst.io/javascript-map-vs-foreach-f38111822c0f"
  }
];

let mockAnswer = "The answer is";
let savedAnswers = [
  {
    answer: mockAnswer,
    question: {
      category: "javascipt",
      link: "https://hacks.mozilla.org/2015/05/es6-in-depth-destructuring/",
      question: "What is object and array destructuring?",
      video: "https://www.youtube.com/watch?v=RJaRRS27100"
    }
  }
];

let updateSavedAnswersMock = jest.fn();
let showSavedAnswersMock = jest.fn();


describe('Question', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Question 
        techQuestions={techQuestions}
        savedAnswers={savedAnswers}
        updateSavedAnswers={updateSavedAnswersMock}
        showSavedAnswers={showSavedAnswersMock}/>
    )
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when updateCurrentAnswer is invoked on change', () => {
    const mockEvent = { target: { value: "My answer is"} }
    wrapper.instance().updateCurrentAnswer(mockEvent);
    expect(wrapper.state().currentAnswer).toEqual("My answer is")
  })

  it('should increment the current index when changeToNextQuestion', () => {
    expect(wrapper.state().currentQuestionIndex).toEqual(0);
    wrapper.instance().changeToNextQuestion();
    expect(wrapper.state().currentQuestionIndex).toEqual(1);
  })

  it('should decrement the current index when changeBackToPreviousQuestion', () => {
    expect(wrapper.state().currentQuestionIndex).toEqual(0);
    wrapper.instance().changeBackToPreviousQuestion();
    expect(wrapper.state().currentQuestionIndex).toEqual(19);
  })


});