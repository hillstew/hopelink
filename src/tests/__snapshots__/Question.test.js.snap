// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question should match snapshot 1`] = `
<AnswerCard
  currentAnswer="The answer is"
  currentQuestion="What is object and array destructuring?"
  currentQuestionIndex={0}
  savedAnswers={
    Array [
      Object {
        "answer": "The answer is",
        "question": Object {
          "category": "javascipt",
          "link": "https://hacks.mozilla.org/2015/05/es6-in-depth-destructuring/",
          "question": "What is object and array destructuring?",
          "video": "https://www.youtube.com/watch?v=RJaRRS27100",
        },
      },
    ]
  }
  showSavedAnswers={[MockFunction]}
  techQuestions={
    Array [
      Object {
        "category": "javascript",
        "link": "https://www.sitepoint.com/demystifying-javascript-variable-scope-hoisting/",
        "question": "What is the difference between local and global variables, and how do they relate to hoisting?",
        "video": "https://www.youtube.com/watch?v=iJKkZA215tQ&t=244s",
      },
      Object {
        "category": "javascript",
        "link": "https://medium.com/javascript-scene/master-the-javascript-interview-what-is-a-closure-b2f0d2152b36",
        "question": "What is closure?",
        "video": "https://www.youtube.com/watch?v=-xqJo5VRP4A",
      },
      Object {
        "category": "javascript",
        "link": "https://tylermcginnis.com/javascript-private-and-public-class-fields/",
        "question": "Describe private and public class fields",
        "video": "https://www.youtube.com/watch?v=ZxpdR_C7RVA",
      },
      Object {
        "category": "javascipt",
        "link": "https://medium.com/@bretdoucette/foreach-v-map-javascript-the-definition-and-the-difference-5eaa65e15b65",
        "question": "What is the difference between forEach and map?",
        "video": "https://codeburst.io/javascript-map-vs-foreach-f38111822c0f",
      },
    ]
  }
  updateSavedAnswers={[MockFunction]}
/>
`;
